pkg/kademlia/peer_discovery.go

```diff
@@ -0,0 +1,126 @@
+// Copyright (C) 2018 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package kademlia
+
+import (
+	"context"
+	"math/big"
+	"sync"
+
+	"go.uber.org/zap"
+
+	"storj.io/storj/pkg/dht"
+	"storj.io/storj/pkg/node"
+	"storj.io/storj/pkg/pb"
+)
+
+type peerDiscovery struct {
+	client node.Client
+	target dht.NodeID
+	opts   discoveryOptions
+
+	cond      sync.Cond
+	queue     *XorQueue
+	contacted map[string]int
+}
+
+func newPeerDiscovery(nodes []*pb.Node, client node.Client, target dht.NodeID, opts discoveryOptions) *peerDiscovery {
+	queue := NewXorQueue(opts.concurrency)
+	queue.Insert(target, nodes)
+
+	return &peerDiscovery{
+		client: client,
+		target: target,
+		opts:   opts,
+
+		cond:      sync.Cond{L: &sync.Mutex{}},
+		queue:     queue,
+		contacted: map[string]int{},
+	}
+}
+
+func (lookup *peerDiscovery) Run(ctx context.Context) error {
+	wg := sync.WaitGroup{}
+
+	// protected by `lookup.cond.L`
+	working := 0
+	allDone := false
+
+	wg.Add(lookup.opts.concurrency)
+	for i := 0; i < lookup.opts.concurrency; i++ {
+		go func() {
+			defer wg.Done()
+			for {
+				var (
+					next     *pb.Node
+					priority big.Int
+				)
+
+				lookup.cond.L.Lock()
+				for {
+					// everything is done, this routine can return
+					if allDone {
+						lookup.cond.L.Unlock()
+						return
+					}
+
+					next, priority = lookup.queue.Closest()
+					if !lookup.opts.bootstrap && priority.Cmp(&big.Int{}) == 0 {
```

Comment:
don't you want to compare that the node being returned and the node you are looking for equal ?

Reply:
Also, I just realized that we might need to connect to the node, otherwise it might not be added to the cache? i.e. let's disable early termination for now, just in case.

Reply:
When the queue is empty it returns `priority = big.Int{}`. Maybe:

```
if next != nil {
	if !lookup.opts.bootstrap && priority.Cmp(&big.Int{}) == 0 {
		allDone = true
	}
	working++
	break
}
```